File: packages/engine/src/cpus/6502/instructions/pha.js
import { CycleType } from '..'

/**
 * PusH Accumulator (in the stack)
 *
 * @example
 * MODE          SYNTAX        HEX LEN TIM
 * Implied       PHA           $48  1   3
 */
export function* pha() {
  this.fetch()
  if (--this.pc < 0) this.pc = 0xffff
  yield CycleType.GET

  this.write(0x100 | this.sp, this.acc)
  if (--this.sp < 0) this.sp = 0xff
  yield CycleType.PUT
}

File: packages/engine/src/cores/famibob/cpu/instructions/pha.zig
const std = @import("std");
const CPU = @import("../cpu.zig").CPU;
const Addressing = @import("../opcode.zig").Addressing;

// PusH Accumulator (in the stack)
//
// MODE          SYNTAX        HEX LEN TIM
// Implied       PHA           $48  1   3

pub fn pha(self: *CPU, addressing: Addressing) void {
    _ = addressing;
    if (self.cycle_counter == 2) {
        _ = self.fetch();
        self.pc -%= 1;
        self.next_cycle = .write;
        return;
    } else {
        self.write(0x100 | @as(u16, self.sp), self.acc);
        self.sp -%= 1;
    }
}
